File: solutions/go/vehicle-purchase/1/vehicle_purchase.go
package purchase

// RequiresLicense is an arary of the types of vehicles that require a license to operate.
var RequiresLicense = []string{
	"car",
	"truck",
}

// Contains checks if a given array of strings contains a particular string.
func Contains(array []string, body string) bool {
	for _, element := range array {
		if element == body {
			return true
		}
	}

	return false
}

// NeedsLicense determines whether a license is needed to drive a type of vehicle. Only "car" and "truck" require a license.
func NeedsLicense(kind string) bool {
	return Contains(RequiresLicense, kind)
}

// ChooseVehicle recommends a vehicle for selection. It always recommends the vehicle that comes first in lexicographical order.
func ChooseVehicle(option1, option2 string) string {
	var option string

	if option1 <= option2 {
		option = option1
	} else {
		option = option2
	}

	return option + " is clearly the better choice."
}

// CalculateResellPrice calculates how much a vehicle can resell for at a certain age.
func CalculateResellPrice(originalPrice, age float64) float64 {
	if age < 3 {
		return originalPrice * 0.8
	}

	if age < 10 {
		return originalPrice * 0.7
	}

	return originalPrice * 0.5
}

File: solutions/javascript/vehicle-purchase/1/vehicle-purchase.js
// @ts-check

/** Types of vehicles that require a license. */
const VEHICLES_REQ_LICENSE = [
  'car',
  'truck'
];

/**
 * Determines whether or not you need a licence to operate a certain kind of vehicle.
 *
 * @param {string} type
 * @returns {boolean} whether a license is required
 */
export function needsLicense(type) {
  return VEHICLES_REQ_LICENSE.includes(type);
}

/**
 * Helps choosing between two options by recommending the one that
 * comes first in dictionary order.
 *
 * @param {string} option1
 * @param {string} option2
 * @returns {string} a sentence of advice which option to choose
 */
export function chooseVehicle(option1, option2) {
  const rec = (option1 < option2) ? option1 : option2;
  return rec + ' is clearly the better choice.';
}

/**
 * Calculates an estimate for the price of a used vehicle in the dealership
 * based on the original price and the age of the vehicle.
 *
 * @param {number} originalPrice
 * @param {number} age
 * @returns {number} expected resell price in the dealership
 */
export function calculateResellPrice(originalPrice, age) {
  if (age > 10) {
    return originalPrice * 0.5;
  }

  if (age >= 3) {
    return originalPrice * 0.7;
  }

  return originalPrice * 0.8;
}
